lib/posix-vfs/posix-vfsroot/vfsroot.c

```diff
@@ -0,0 +1,231 @@
+		case '.':
+			if (cur + 1 < len)
+				switch (path[cur + 1]) {
+				case '/':
+					cur += 2;
+					continue;
+				case '.':
+					if (cur + 2 == len) {
+						cur += 2;
+						break;
+					}
+					if (cur + 2 < len &&
+					    path[cur + 2] == '/') {
+						cur += 3;
+						continue;
+					}
+				}
+		}
+		break;
+	}
+
+	if (!(flags & UKFS_LOOKUP_IGNMNT)) {
+		uk_spin_lock(&umount_lock);
+		mnt = rootmnt;
+		if (mnt)
+			uk_file_acquire(mnt);
+		uk_spin_unlock(&umount_lock);
+	}
+
+	if (mnt) {
+		if (!(flags & UKFS_LOOKUP_NO_MNTAUX)) {
+			uk_file_acquire(f);
+			out->aux = f;
+		}
+		out->target = mnt;
+		*nout = cur;
+		return UKFS_STOP_MNT;
+	}
+	if (cur == len) {
+		uk_file_acquire(f);
+		out->target = f;
+		return UKFS_SUCCESS;
+	}
+	return -ENOENT;
+}
+
+static
+int root_mount(const struct uk_file *f __maybe_unused,
+	       const struct uk_file **target)
+{
+	const struct uk_file *t = *target;
+
+	UK_ASSERT(f == &uk_posix_vfsroot_file);
+	if (t) {
+		const struct uk_file *prev = NULL;
+
+		if (likely(uk_compare_exchange_n(&rootmnt, &prev, t)))
+			uk_file_acquire(t);
+		else
+			return -EEXIST;
+	} else {
+		uk_spin_lock(&umount_lock);
+		t = uk_exchange_n(&rootmnt, NULL);
+		uk_spin_unlock(&umount_lock);
+		if (likely(t))
+			*target = t;
+		else
+			return -ENOENT;
+	}
+	return 0;
+}
+
+static
+ssize_t root_listdir(const struct uk_file *f __maybe_unused, size_t *curp,
+		     void *buf, size_t len)
+{
+	struct uk_fs_dirent *out = buf;
+	size_t p = *curp;
+	size_t rlen = 0;
+	size_t sz;
+
+	UK_ASSERT(f == &uk_posix_vfsroot_file);
+	/* Output '.' & '..' dentries pointing to self */
+	while (p < 2) {
+		sz = UKFS_DIRENT_RECLEN(p ? 2 : 1);
+		if (unlikely(sz > len)) {
+			if (rlen)
+				break;
+			return -EINVAL;
+		}
+		if (!p)
+			UKFS_DIRENT_OUT_DOT(out, 0);
+		else
+			UKFS_DIRENT_OUT_DOTDOT(out, 0);
+		rlen += sz;
+		len -= sz;
+		out = UKFS_DIRENT_NEXT(out, sz);
+		p++;
+	}
+	*curp = p;
+	return rlen;
+}
+
+static const struct uk_fs_ops rootfsops = {
+	.stat = root_fsstat,
+	.sync = uk_fs_common_nop_sync,
+	.lookup = root_lookup,
+	.mount = root_mount,
+	.rebind = uk_fs_common_nop_rebind,
+	.listdir = root_listdir,
+	.create = uk_fs_common_rofs_create,
+	.unlink = uk_fs_common_rofs_unlink,
+	.rename = uk_fs_common_rofs_rename,
+	.graft = uk_fs_common_nop_graft,
+	.readlink = NULL,
+	.constraints = UKFS_MODE_NOIOLOCK
+};
+
+static
+struct uk_file_state root_fstate = UK_FILE_STATE_EVENTS_INITIALIZER(root_fstate,
+	UKFD_POLLIN | UKFD_POLLOUT);
+
+/* uk_posix_vfsroot_file starts off with a refcount of 3:
+ * - this static reference
+ * - reference in uk_posix_vfsroot_ofile
+ * - .fsroot reference in initial VFS state
+ */
+static uk_file_refcnt root_refcnt = UK_FILE_REFCNT_INITIALIZER_V(root_refcnt,
+								 3);
+
+const struct uk_file uk_posix_vfsroot_file = {
+	.vol = ROOT_VOLID,
+	.node = NULL,
+	.ops = &rootfops,
+	.fsops = &rootfsops,
+	.refcnt = &root_refcnt,
+	.state = &root_fstate,
+	._release = uk_file_static_release,
```

Comment:
I can't help but notice that in some places you do add `,` for the last statically initialized struct member but in other places you don't... :sweat_smile: